src/coreclr/gc/gc.cpp

```diff
@@ -23578,8 +24312,677 @@ void gc_heap::equalize_promoted_bytes()
+
+                generation_start_segment (gen) = nullptr;
+                generation_tail_ro_region (gen) = nullptr;
+                generation_tail_region (gen) = nullptr;
+            }
+        }
+    }
+
+    // transfer the free regions from the heaps going idle
+    for (int i = new_n_heaps; i < n_heaps; i++)
+    {
+        gc_heap* hp = g_heaps[i];
+        int dest_heap_number = i % new_n_heaps;
+        gc_heap* hpd = g_heaps[dest_heap_number];
+
+        for (int kind = 0; kind < count_free_region_kinds; kind++)
+        {
+            hpd->free_regions[kind].transfer_regions(&hp->free_regions[kind]);
+        }
+    }
+    // update number of heaps
+    n_heaps = new_n_heaps;
+    gc_t_join.update_n_threads(new_n_heaps);
+#ifdef BACKGROUND_GC
+    bgc_t_join.update_n_threads(new_n_heaps);
+#endif //BACKGROUND_GC
+
+    // even out the regions over the current number of heaps
+    equalize_promoted_bytes (max_generation);
+
+    total_num_fl_items_moved_stage1 = ~0;
+    total_num_fl_items_stage1 = ~0;
+
+    // adjust data for the heaps now in operation
+    for (int i = 0; i < new_n_heaps; i++)
+    {
+        gc_heap* hp = g_heaps[i];
+        generation* gen0 = hp->generation_of (0);
+        if ((hp->ephemeral_heap_segment == nullptr) ||
+            (heap_segment_heap (hp->ephemeral_heap_segment) != hp))
+        {
+            hp->ephemeral_heap_segment = heap_segment_rw (generation_start_segment (gen0));
+            hp->alloc_allocated = heap_segment_allocated (hp->ephemeral_heap_segment);
+        }
+
+        for (int gen_idx = 0; gen_idx < total_generation_count; gen_idx++)
+        {
+            generation* gen = hp->generation_of (gen_idx);
+            heap_segment *allocation_region = generation_allocation_segment (gen);
+            if ((allocation_region == nullptr) ||
+                (heap_segment_heap (allocation_region) != hp))
+            {
+                generation_allocation_segment (gen) = heap_segment_rw (generation_start_segment (gen));
+            }
+        }
+    }
+
+    // rethread the free lists
+    for (int gen_idx = 0; gen_idx < total_generation_count; gen_idx++)
+    {
+        for (int i = 0; i < old_n_heaps; i++)
+        {
+            gc_heap* hp = g_heaps[i];
+
+            hp->rethread_fl_items (gen_idx);
+        }
+        merge_fl_from_other_heaps (gen_idx, new_n_heaps, old_n_heaps);
+    }
+
+    // there should be no items in the bgc_alloc_lock on any heap - check that
+    for (int i = 0; i < old_n_heaps; i++)
+    {
+        gc_heap* hp = g_heaps[i];
+
+        hp->bgc_alloc_lock->check();
+    }
+
+    // put heaps that going idle now into the decommissioned state
+    for (int i = n_heaps; i < old_n_heaps; i++)
+    {
+        gc_heap* hp = g_heaps[i];
+
+        hp->decommission_heap();
+    }
+
+    // make sure no allocation contexts point to idle heaps
+    fix_allocation_contexts_heaps();
+
+    GCConfig::s_HeapVerifyLevel = 1;
```

Comment:
you could just call verify_heap here instead of turning it on for all GCs later as we discussed (or verify this one and a couple of GCs later

Reply:
Hmm, this code is on a single thread, so I'd have to iterate over all heaps. Needs more thought.

Reply:
ahh right, one idea is to set a global var to indicate that we did change the HC (heap count) like -

`PER_HEAP_ISOLATED_FIELD_SINGLE_GC bool last_suspension_hc_changed_p; `

and next time we do a GC, we will call `verify_heap` on entry if this is true and that GC will reset it. or you could keep a var to record the GC index when we suspended to change the HC, and call `verify_heap` in the next few GCs if this is not -1.

Reply:
Actually, in the dynamic heap count PR building on this one, I execute the `change_heap_count` method on all GC threads, and so it would be easy to just call `verify_heap` at the end.